### ISSUE TITLE
Conditional wait

### ISSUE THREAD

**Issue by AngryMane** [capnproto/capnproto#1692]
Hi! I'm developping a Cap'n Proto client app. 
This app need to do some finalize process when users hult the app by SIGINT(ctrl+c), but on the other hand, the app needs to wait callback from server. To resolve this, I want to do  somewhat like `NEVER_DONE.wait(wait_scope, predicate)` , as same as
[std::condition_variable::wait](https://cpprefjp.github.io/reference/condition_variable/condition_variable/wait.html) so that the app end to wait event loop when hulting.
Simply put, I want to wait until a specific condition is met.
Is there any way to do this?

**Comment by Sarcasm** [comment]
I use exclusive join and then()-continuation for this:

```cpp  
  auto onSignal = io.unixEventPort.onSignal(SIGINT)
                        .exclusiveJoin(io.unixEventPort.onSignal(SIGTERM))
                        .then([](siginfo_t&& sig) {
                            return TerminationStatus::signal(sig.si_signo);
                        });

  auto promise = ...

   return promise.exclusiveJoin(kj::mv(onSignal))
        .then([](const TerminationStatus& status) {
            return finalizeProcess().then([status]() { return status;  });
        })
        .wait(io.waitScope);
```

You may consider adding a timeout to the finalizeProcess() step so that the user isn't deceived after a Ctrl+C/SIGTERM.
Clean shutdown is nice to have but there is also the possibility an uncatchable signal was sent, e.g. SIGQUIT, ideally the application should be robust to such abnormal exits too.

**Comment by AngryMane** [comment]
Thank you for your reply!  I was in a lot of trouble so this is very helpful. I'll try it! 😊

**Comment by AngryMane** [comment]
If I use EzRpcServer or EzRpcClient, is there any way to get the unixEventPort?

EzRpcServer creates unixEventPort instance and intert it to eventloop at setupAsyncIo() defined in async-io-unix.c++. (I'm working on WSL)
In this case, unixEventPort instance is hold at following locations. 

* LowLevelAsyncIoProviderImpl

    I can access LowLevelAsyncIoProviderImpl instance by using EzRpcServer::getLowLevelIoProvider().
    But I can't access unixEventPort instance by following reasons.
        * LowLevelAsyncIoProviderImpl is defined in async-io-unix.c++, not in async-io.h, so I can't refer to the declaration.
        * unixEventPort instance is private member of LowLevelAsyncIoProviderImpl. 
 
* EzRpcContext

    I can access EzRpcContext instance by using EzRpcContext::getThreadLocal().
    But unixEventPort instance is private member and there is no method to get it. 


I think EzRpcContext should provide a method to get it...

**Comment by AngryMane** [comment]
[kj::LowLevelAsyncIoProvider](https://github.com/capnproto/capnproto/blob/79dab94b2d4b73247c4e68cc8a31fa73b39dfbda/c%2B%2B/src/kj/async-io.h#L769) does not have a member named unixEventPort.
[kj::LowLevelAsyncIoProviderImpl](https://github.com/capnproto/capnproto/blob/79dab94b2d4b73247c4e68cc8a31fa73b39dfbda/c%2B%2B/src/kj/async-io-unix.c%2B%2B#L1475) is an implementation of kj::LowLevelAsyncIoProvider.
This class owns a member named eventPort as a private member and provides access to it with the function getEventPort(), but since kj::LowLevelAsyncIoProvider does not declare getEventPort() I can't use this.
I also checked other functions provided by kj::LowLevelAsyncIoProvider, and wrapInputFd() was suspect, but it still does not seem to be able to achieve what I want to do this time.

**Comment by Sarcasm** [comment]
Oops sorry, in the link I pasted I did not notice I was into the `AsyncIoContext` struct and no longer in the ` AsyncIoProvider` class.

I think you are right, and the EzRpcClient interface talks about the limitations:

https://github.com/capnproto/capnproto/blob/79dab94b2d4b73247c4e68cc8a31fa73b39dfbda/c%2B%2B/src/capnp/ez-rpc.h#L87-L92

**Comment by AngryMane** [comment]
@Sarcasm 
I see. I tried to create a RPC wappre class base on EzRPCServer, and it works now! 
https://github.com/AngryMane/cap-sample/blob/main/server/custom/custom-rpc.h

I just added getUnixEventPort() method to both CustomRpcContext(EzRpcContext) and CustomRpcServer(EzRpcServer).

thank you for your kind answer! 😊

**Comment by kentonv** [comment]
FWIW I plan to deprecate the "EZ" interfaces, I think they were a mistake for exactly these sorts of reasons. Setting up the I/O context using `kj::setupAsyncIo()` and `capnp::TwoPartyServer`/`capnp::TwoPartyClient` is not too much harder and provides a lot more flexibility.

**Comment by AngryMane** [comment]
I created sample code for using TwoPartyServer and TwoPartyClient. 
https://github.com/AngryMane/cap-sample/blob/main/
I share this because someone may need it when they come to this issue.

**Comment by kentonv** [comment]
Thanks. Yeah, obviously I can't really deprecate the EZ interfaces while still having the docs recommending them... will need to update those at some point.
